### PR TITLE
Taxonomy bigdatext devas 1510

### PR DESCRIPTION
**Pull request by amensouissi**

### CODE REVIEW

File: assembl/graphql/post.py [hunk 1]
```diff
@@ -96,6 +97,9 @@ class Meta:
     attachments = graphene.List(PostAttachment)
     original_locale = graphene.String()
     publishes_synthesis = graphene.Field(lambda: Synthesis)
+    type = graphene.String()
+    discussion_id = graphene.String()
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/post.py]: an external app like Bigdatext doesn't have access to this data

File: assembl/graphql/discussion.py [hunk 2]
```diff
@@ -32,6 +32,7 @@ class Meta:
     button_label_entries = graphene.List(LangStringEntry)
     header_image = graphene.Field(Document)
     logo_image = graphene.Field(Document)
+    slug = graphene.String()
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/discussion.py]: an external app like Bigdatext doesn't have access to this data

File: assembl/graphql/extract.py [hunk 3]
```diff
@@ -37,11 +43,15 @@ class Meta:
     creation_date = DateTime()
     creator_id = graphene.Int()
     creator = graphene.Field(lambda: AgentProfile)
+    extract_state = graphene.Field(type=ExtractStates)
 
     def resolve_creator(self, args, context, info):
         if self.creator_id is not None:
             return models.AgentProfile.get(self.creator_id)
 
+    def resolve_extract_state(self, args, context, info):
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/extract.py]: effectivement

File: assembl/graphql/idea.py [hunk 4]
```diff
@@ -185,6 +185,7 @@ class Meta:
     message_columns = graphene.List(lambda: IdeaMessageColumn)
     ancestors = graphene.List(graphene.ID)
     vote_results = graphene.Field(VoteResults, required=True)
+    type = graphene.String()
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/idea.py]: this is an oversight.

File: assembl/graphql/post.py [hunk 5]
```diff
@@ -330,13 +357,16 @@ def mutate(root, args, context, info):
...
             new_post = cls(
                 discussion=discussion,
                 subject=subject_langstring,
                 body=body_langstring,
                 creator_id=user_id,
                 body_mime_type=u'text/html',
-                message_classifier=classifier
+                message_classifier=classifier,
+                creation_date=now,
+                modification_date=now,
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/post.py]: In my opinion, the creation date is the same as the date of modification for a new post (the same logic is used in the file systems for example).

"Semantically, this could be wrong, as querying for a modified post would now be completely skewed by this change.". Please see the `modified` attr. I think that the post is modified if the modification date is greater than the creation date.

However, I agree with you on adding a getter for the modification_date attribute (and not a new attribute).

File: assembl/graphql/post.py [hunk 6]
```diff
@@ -666,3 +696,54 @@ def mutate(root, args, context, info):
...
+        discussion_id = context.matchdict['discussion_id']
+        # Retrieve the user id
+        user_id = context.authenticated_userid or Everyone
+        extracts = args.get('extracts')
+        status = True
+        # Add all of extracts
+        for extract in extracts:
+            post_id = extract.get('post_id')
+            post_id = int(Node.from_global_id(post_id)[1])
+            post = models.Post.get(post_id)
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/post.py]: Effectivement :)

File: assembl/graphql/extract.py [hunk 7]
```diff
@@ -96,3 +106,25 @@ def mutate(root, args, context, info):
...
+    def mutate(root, args, context, info):
+        extract_id = args.get('extract_id')
+        extract_id = int(Node.from_global_id(extract_id)[1])
+        extract = models.Extract.get(extract_id)
+        require_instance_permission(CrudPermissions.UPDATE, extract, context)
+        user_id = context.authenticated_userid or Everyone
+        # Publish the extract
+        extract.extract_state = models.ExtractStates.PUBLISHED.value
+        extract.owner_id = user_id
+        extract.creator_id = user_id
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/extract.py]: Currently, for the Extracts the user shown is the creator and not the owner. For the published Extracts submitted by the Robot, the moderator must be displayed. In this case we need to display the owner and not the creator in the front-end. @fred

File: assembl/static2/js/app/components/harvesting/harvestingBox.jsx [hunk 8]
```diff
@@ -239,6 +276,23 @@ class DumbHarvestingBox extends React.Component<Props, State> {
       });
   };
 
+  confirmHarvesting = (): void => {
+    const { extract, confirmExtract, refetchPost } = this.props;
+    const variables = {
+      extractId: extract ? extract.id : null
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/harvesting/harvestingBox.jsx]: Yes

File: assembl/static2/js/app/components/harvesting/harvestingBox.jsx [hunk 9]
```diff
@@ -290,10 +363,14 @@ class DumbHarvestingBox extends React.Component<Props, State> {
...
       extract && extract.creator && extract.creator.displayName ? extract.creator.displayName : getConnectedUserName();
+    const extractState = extract && extract.extractState;
+    const isSubmitted = extractState === ExtractStates.SUBMITTED;
     const userName = isAuthorAccountDeleted ? I18n.t('deletedUser') : harvesterUserName;
     const harvesterUserId = extract && extract.creator && extract.creator.userId ? extract.creator.userId : getConnectedUserId();
+    const menuDisabled = disabled || isSubmitted;
+    const hasFooter = disabled || isEditable || isSubmitted;
     return (
-      <div>
+      <div className={isSubmitted ? 'submitted-harvesting' : ''}>
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/harvesting/harvestingBox.jsx]: That's enough I think

File: assembl/tests/views_tests/test_graphql.py [hunk 10]
```diff
@@ -2829,3 +2829,45 @@ def __init__(self, filename, type):
     assert res_discussion['titleEntries'][0]['value'] == u'My title'
     assert res_discussion['subtitleEntries'][0]['value'] == u'My subtitle'
     assert res_discussion['buttonLabelEntries'][0]['value'] == u'My button label'
+
+
+def test_get_all_posts(graphql_request, proposition_id):
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/tests/views_tests/test_graphql.py]: Yes

File: assembl/views/__init__.py [hunk 11]
```diff
@@ -214,7 +214,7 @@ def get_default_context(request, **kwargs):
                        'localhost'))
         react_url = 'http://%s:%d' % (
             webpack_host,
-            int(config.get('webpack_port', 8080)))
+            int(config.get('webpack_port', 8000)))
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/views/__init__.py]: Oh yes

File: assembl/views/api2/discussion.py [hunk 12]
```diff
@@ -1614,6 +1618,21 @@ def update_notification_subscriptions(request):
...
+discussions_slugs = Service(
+    name='discussions_slugs',
+    path='/discussions_slugs',
+    description="List of existing Discussion slugs",
+    renderer='json'
+)
+
+
+@discussions_slugs.get()
+def get_discussions_slugs(request):
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/views/api2/discussion.py]: See the next line, I use `discussions_with_access(userid, permission=P_READ)` method

File: assembl/views/api2/discussion.py [hunk 13]
```diff
@@ -1614,6 +1618,21 @@ def update_notification_subscriptions(request):
     return {'status': 'Notification subscriptions have been updated.'}
 
 
+discussions_slugs = Service(
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/views/api2/discussion.py]: but it's used only by the V2. Is it relevant to move it?

File: assembl/graphql/post.py [hunk 14]
```diff
@@ -666,3 +696,54 @@ def mutate(root, args, context, info):
...
 
         return AddPostExtract(post=post)
+
+
+# Used by the Bigdatext app
+class AddPostsExtract(graphene.Mutation):
+    class Input:
+        extracts = graphene.List(PostExtractEntryInput, required=True)
+        extract_nature = graphene.String()
+        extract_state = graphene.String()
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/post.py]: Yes, I think we need a Graphene Enum here. And I think we also need to change the UpdateExtract mutation.

File: fabfile.py [hunk 15]
```diff
@@ -592,6 +593,15 @@ def build_virtualenv():
         return
     run('python2 -mvirtualenv --no-setuptools %(venvpath)s' % env)
     # create the virtualenv with --no-setuptools to avoid downgrading setuptools that may fail
+    if not exists("%(projectpath)s/../bluenove-actionable/" % env):
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply fabfile.py]: I added a configuration key '_uses_bluenove_actionable' to indicate its existence.
I spoke with Fred and he told me that it's necessary that Bigdatext App must be installed on the same server as the Assembl App.

File: assembl/graphql/post.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vincentfretin) [reply assembl/graphql/post.py]: You shouldn't do it explicitly here. creation_date is defined like this on the `Content` class
`creation_date = Column(DateTime, nullable=False, default=datetime.utcnow)`

You added `modification_date = Column(DateTime)` to `AssemblPost. I think you should put 
`modification_date = Column(DateTime, nullable=False, default=datetime.utcnow)`
to the `Content` class instead (models/generic.py) instead, and remove this change here.

[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/post.py]: Please see the latest changes :)

[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/graphql/post.py]: it doesn't solve the problem. There will always be a difference between the dates.

File: assembl/graphql/schema.py [hunk 16]
```diff
@@ -116,6 +119,11 @@ class Query(graphene.ObjectType):
     text_fields = graphene.List(ConfigurableFieldUnion, description=docs.Schema.text_fields)
     profile_fields = graphene.List(ProfileField, description=docs.Schema.profile_fields)
     timeline = graphene.List(DiscussionPhase, description=docs.Schema.timeline)
+    posts = SQLAlchemyConnectionField(
+        PostConnection,
+        start_date=graphene.String(description=docs.Schema.posts.start_date),
+        end_date=graphene.String(description=docs.Schema.posts.start_date),
```
Comment (vincentfretin) [inline assembl/graphql/schema.py]: wrong docstring, should be end_date

File: assembl/models/auth.py [hunk 17]
```diff
@@ -685,6 +690,40 @@ def __init__(self, **kwargs):
...
+        machines_data = default_config.get('machine.machines', '')
+        if machines_data:
+            from ..auth.util import add_user
+            db = db or cls.default_db()
+            # retrieve existing machines
+            machines = db.query(cls).filter(
+                cls.is_machine == True  # noqa: E712
+                )
+            # retrieve machines ids
+            all_machines_ids = [machine.username.username for machine in machines]
```
Comment (vincentfretin) [inline assembl/models/auth.py]: Please use the `username_p` property here. We may remove the Username class in the near future.

File: assembl/models/auth.py [hunk 18]
```diff
@@ -685,6 +690,40 @@ def __init__(self, **kwargs):
...
+                    add_user(
+                        machine.get('name'),
+                        None,
+                        machine.get('password'),
+                        R_CATCHER,
+                        username=machine_id,
+                        is_machine=True
+                    )
+                except Exception:
+                    continue
```
Comment (vincentfretin) [inline assembl/models/auth.py]: You catch the exception in the case the user already exists? This may hide other errors... You should check if the user exists first. With `session.query(Username).filter_by(username=machine_id).count())` for example.

File: assembl/scripts/ini_files.py [hunk 19]
```diff
@@ -177,6 +177,7 @@ def generate_ini_files(config, config_fname):
             'autostart_webpack',
             'autostart_elasticsearch_server',
             'autostart_urlmetadata',
+            'autostart_bluenove_actionable',
```
Comment (vincentfretin) [inline assembl/scripts/ini_files.py]: I saw recently there is a list around line 62 with default values, can you please add autostart_urlmetadata and autostart_bluenove_actionable there? Thanks.

Comment (amensouissi) [reply assembl/scripts/ini_files.py]: old code ;)

File: assembl/graphql/utils.py [hunk 20]
```diff
@@ -189,3 +192,46 @@ def update_attachment(discussion, attachment_model, new_value, attachments, atta
...
+
+    if identifiers_with_posts:
+        is_multi_columns = Phases.multiColumns.value in identifiers_with_posts and \
+            len(identifiers_with_posts) == 1
+        ideas.extend(
+            get_ideas(
+                discussion.id,
+                Phases.multiColumns.value if is_multi_columns else None
+            ).all()
+        )
```
Comment (vincentfretin) [inline assembl/graphql/utils.py]: If you call `get_posts_for_phases(discussion, ['thread', 'multiColumns']` you will have duplicate ideas because ideas from the multiColumns phase are actually all the ideas of the thread phase that have multi columns enabled. You should use a set, not a list to remove duplicate ideas here.

File: assembl/graphql/utils.py [hunk 21]
```diff
@@ -189,3 +192,46 @@ def update_attachment(discussion, attachment_model, new_value, attachments, atta
...
+        return None
+
+    model = models.AssemblPost
+    query = discussion.db.query(model)
+    query_source = query
+    for index, idea in enumerate(ideas):
+        related = idea.get_related_posts_query(True)
+        related_query = query_source.join(
+            related, model.id == related.c.post_id
+        )
```
Comment (vincentfretin) [inline assembl/graphql/utils.py]: You currently include deleted posts here. You may want to add an option to filter on `publication_state == models.PublicationStates.PUBLISHED` (see `get_published_posts`)

Comment (amensouissi) [reply assembl/graphql/utils.py]: Please see get_related_posts_query => include_deleted=False

Comment (vincentfretin) [reply assembl/graphql/utils.py]: include_deleted is False by default, right, but it's actually not sufficient. This include_deleted=False means adding the filter `query.filter(content.tombstone_date == None)` but we don't tombstone delete posts having a parent anymore, so you need an additional filter on publication_state.

File: assembl/static2/js/app/utils/translations.js [hunk 22]
```diff
@@ -23,9 +23,13 @@ const Translations = {
       harvesting: "Attrapage",
       inProgress: "Attrapage en cours",
       validated: "Attrapage validé",
+      confirm: "Confirmer",
+      reject: "Rejeter",
       now: "Maintenant",
       harvestingSuccess: "L'extrait a été mis à jour avec succès.",
       harvestingDeleted: "L'extrait a été supprimé avec succès.",
+      harvestingConfirmed: "L'extrait a été confirmé avec succès.",
+      harvestingSubmitted: "L'attrapage est en attente de validation.",
```
Comment (vincentfretin) [inline assembl/static2/js/app/utils/translations.js]: On a `extrait` parout, pourquoi `attrapage` ici ?

Comment (amensouissi) [reply assembl/static2/js/app/utils/translations.js]: effectivement

File: assembl/tests/fixtures/idea_content_links.py [hunk 23]
```diff
@@ -183,6 +183,50 @@ def fin():
...
+        xpath_start=xpathStart,
+        offset_start=offsetStart,
+        xpath_end=xpathEnd,
+        offset_end=offsetEnd
+    )
+    test_session.add(new_range)
+    test_session.flush()
+
+    def fin():
+        print "finalizer extract_with_range_in_reply_post_1"
```
Comment (vincentfretin) [inline assembl/tests/fixtures/idea_content_links.py]: The name should match the fixture name.

File: fabfile.py [hunk 24]
```diff
@@ -844,6 +848,111 @@ def updatemaincode(backup=False):
...
+@task
+def install_bluenove_actionable():
+    """
+    Install the bluenove_actionable app.
+    """
+    if not exists("%(projectpath)s/../bluenove-actionable/" % env):
+        print cyan("Cloning git bluenove-actionable repository")
+        with cd("%(projectpath)s/.." % env):
+            # We need an ssh access
+            run('git clone git://github.com/bluenove/bluenove-actionable.git')
```
Comment (vincentfretin) [inline fabfile.py]: This is an anonymous url, you will never access the private repository with this. You need the ssh version `git@github.com:bluenove/bluenove-actionable.git`

File: fabfile.py [hunk 25]
```diff
@@ -844,6 +848,111 @@ def updatemaincode(backup=False):
...
+
+        with cd("%(projectpath)s/../bluenove-actionable/" % env):
+            run('docker-compose build', warn_only=True)
```
Comment (vincentfretin) [inline fabfile.py]: use `sudo` instead of `run` here.

File: fabfile.py [hunk 26]
```diff
@@ -844,6 +848,111 @@ def updatemaincode(backup=False):
...
+    """
+    Update the bluenove_actionable app. Updating the Git repository and building.
+    """
+    path = join(env.projectpath, '..', 'bluenove-actionable')
+    if exists(path):
+        print(cyan('Updating bluenove-actionable Git repository'))
+        with cd(path):
+            # We need an ssh access
+            run('git pull', warn_only=True)
+            execute(stop_bluenove_actionable)
```
Comment (vincentfretin) [inline fabfile.py]: You don't need to stop when building a new image.

File: fabfile.py [hunk 27]
```diff
@@ -843,6 +853,67 @@ def updatemaincode(backup=False):
...
+            run('docker-compose build --no-cache', warn_only=True)
+
+
+@task
+def stop_bluenove_actionable():
+    path = join(env.projectpath, '..', 'bluenove-actionable')
+    if exists(path):
+        print(cyan('stop bluenove-actionable'))
+        with cd(path):
+            fabsudo('docker kill bluenoveact', warn_only=True)
```
[the remark this answers is not in the document]
Comment (vincentfretin) [reply fabfile.py]: yeah use `sudo` and not `fabsudo` here

[the remark this answers is not in the document]
Comment (vincentfretin) [reply fabfile.py]: `docker kill` is hard, you should try to stop first. Better, use docker-compose too here. `docker-compose down` is what you want, it will stop, kill if necessary and remove the container.

File: fabfile.py [hunk 28]
```diff
@@ -844,6 +848,111 @@ def updatemaincode(backup=False):
...
+
+
+@task
+def start_bluenove_actionable():
+    """
+    Start the bluenove_actionable app.
+    To start the application we need three environment variables:
+    - URL_INSTANCE: The URL of the Assembled Instance.
```
Comment (vincentfretin) [inline fabfile.py]: Assembled Instance => Assembl instance

File: fabfile.py [hunk 29]
```diff
@@ -844,6 +848,111 @@ def updatemaincode(backup=False):
...
+    robot = get_robot_machine()
+    if exists(path) and robot:
+        print(cyan('run bluenove-actionable'))
+        with cd(path):
+            with shell_env(
+                URL_INSTANCE=env.public_hostname,
+                ROBOT_IDENTIFIER=robot.get('identifier'),
+                ROBOT_PASSWORD=robot.get('password')
+            ):
+                fabsudo('docker-compose up -d', warn_only=True)
```
Comment (vincentfretin) [inline fabfile.py]: use `sudo` instead of `fabsudo`

File: fabfile.py (hunk 26 above)
Comment (vincentfretin) [reply fabfile.py]: and you probably want a `restart_bluenove_actionable` just after the `docker-compose build`

File: fabfile.py [hunk 30]
```diff
@@ -844,6 +848,111 @@ def updatemaincode(backup=False):
 
             venvcmd_py3('pip install -r ../url_metadata/requirements.txt')
 
+        if env.uses_bluenove_actionable:
+            execute(update_bluenove_actionable)
```
Comment (vincentfretin) [inline fabfile.py]: We may not want to rebuild the docker image each time we deploy. What time does it take to rebuild the image? I'll remove those lines. @ayyazdaniaryan what do you think?

**Comment by vincentfretin** [comment]
Really good work on this PR. My latest comments are mainly on optimizing the query to get all posts and the fab tasks for the deployment of the container.

File: assembl/graphql/utils.py (hunk 20 above)
Comment (amensouissi) [reply assembl/graphql/utils.py]: Please see get_ideas.

File: assembl/graphql/utils.py (hunk 21 above)
Comment (amensouissi) [reply assembl/graphql/utils.py]: done

File: fabfile.py (hunk 24 above)
Comment (amensouissi) [reply fabfile.py]: effectivement

File: fabfile.py (hunk 30 above)
Comment (amensouissi) [reply fabfile.py]: Yes we can do that manually!

File: assembl/auth/util.py [hunk 31]
```diff
@@ -447,29 +454,33 @@ def add_user(name, email, password, role, force=False, username=None,
...
-            verified=True)
-        db.add(account)
+        else:
+            account = EmailAccount(
+                profile=user,
+                email=email,
+                preferred=True,
+                verified=True)
+            db.add(account)
+
```
Comment (vincentfretin) [inline assembl/auth/util.py]: @amensouissi why did you change this? The else was associated to the for, if no user.accounts, it should execute the else.

**Comment by ayyazdaniaryan** [comment]
Good work @amensouissi ! 👍
